File: src/scpn_phase_orchestrator/coupling/ei_balance.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Phase Orchestrator — Excitatory/Inhibitory balance

"""Excitatory/inhibitory balance summaries and adjustment helpers.

The module measures mean outgoing coupling from caller-specified excitatory and
inhibitory index sets, then optionally rescales inhibitory rows toward a target
ratio. Rust acceleration is used when available; the NumPy fallback preserves
the same shape and summary contract for examples and deterministic tests.
"""

from __future__ import annotations

from dataclasses import dataclass
from numbers import Real
from typing import TypeAlias

import numpy as np
from numpy.typing import NDArray

try:
    from spo_kernel import (
        adjust_ei_ratio_rust as _rust_adjust,
    )
    from spo_kernel import (
        compute_ei_balance_rust as _rust_ei,
    )

    _HAS_RUST = True
except ImportError:
    _HAS_RUST = False

__all__ = ["EIBalance", "compute_ei_balance", "adjust_ei_ratio"]

FloatArray: TypeAlias = NDArray[np.float64]


def _contains_boolean_alias(value: object) -> bool:
    raw = np.asarray(value, dtype=object)
    return any(isinstance(item, bool) for item in raw.ravel())


def _validate_knm(value: object) -> FloatArray:
    if _contains_boolean_alias(value):
        raise ValueError("knm must not contain boolean values")
    try:
        knm = np.asarray(value, dtype=np.float64)
    except (TypeError, ValueError) as exc:
        raise ValueError("knm must be a finite square matrix") from exc
    if knm.ndim != 2 or knm.shape[0] != knm.shape[1]:
        raise ValueError("knm must be a finite square matrix")
    if not np.all(np.isfinite(knm)):
        raise ValueError("knm must contain only finite values")
    return np.ascontiguousarray(knm, dtype=np.float64)


def _validate_target_ratio(value: object) -> float:
    if isinstance(value, bool) or not isinstance(value, Real):
        raise TypeError("target_ratio must be a finite positive real")
    target_ratio = float(value)
    if not np.isfinite(target_ratio) or target_ratio <= 0.0:
        raise ValueError("target_ratio must be a finite positive real")
    return target_ratio


@dataclass
class EIBalance:
    """Summary of excitatory and inhibitory coupling balance."""

    ratio: float
    excitatory_strength: float
    inhibitory_strength: float
    is_balanced: bool


def _validate_indices(indices: list[int], n: int, name: str) -> list[int]:
    valid: list[int] = []
    for idx in indices:
        if not isinstance(idx, int) or isinstance(idx, bool):
            msg = f"{name} indices must be integers, got {idx!r}"
            raise ValueError(msg)
        if idx < 0:
            msg = f"{name} indices must be non-negative, got {idx}"
            raise ValueError(msg)
        if idx < n:
            valid.append(idx)
    return valid


def compute_ei_balance(
    knm: FloatArray,
    excitatory_indices: list[int],
    inhibitory_indices: list[int],
) -> EIBalance:
    """Compute E/I balance from coupling matrix and layer typing.

    Kuroki & Mizuseki 2025, Neural Computation — E/I balance is the
    critical parameter for synchronization, not K or D.

    ratio > 1: excitation-dominated (hypersynchrony risk)
    ratio < 1: inhibition-dominated (desynchronization risk)
    ratio ≈ 1: balanced (optimal for metastability)
    """
    knm = _validate_knm(knm)
    n = knm.shape[0]
    excitatory_indices = _validate_indices(excitatory_indices, n, "excitatory")
    inhibitory_indices = _validate_indices(inhibitory_indices, n, "inhibitory")

    if _HAS_RUST:
        k_flat = np.ascontiguousarray(knm.ravel())
        e_arr = np.array(excitatory_indices, dtype=np.int64)
        i_arr = np.array(inhibitory_indices, dtype=np.int64)
        ratio, e_str, i_str, balanced = _rust_ei(k_flat, n, e_arr, i_arr)
        return EIBalance(
            ratio=float(ratio),
            excitatory_strength=float(e_str),
            inhibitory_strength=float(i_str),
            is_balanced=bool(balanced),
        )

    e_mask = np.zeros(n, dtype=bool)
    i_mask = np.zeros(n, dtype=bool)
    for idx in excitatory_indices:
        e_mask[idx] = True
    for idx in inhibitory_indices:
        i_mask[idx] = True

    # Excitatory strength: mean coupling FROM excitatory oscillators
    e_strength = float(np.mean(knm[e_mask, :])) if np.any(e_mask) else 0.0
    # Inhibitory strength: mean coupling FROM inhibitory oscillators
    i_strength = float(np.mean(knm[i_mask, :])) if np.any(i_mask) else 0.0

    if i_strength < 1e-15:
        ratio = float("inf") if e_strength > 0 else 1.0
    else:
        ratio = e_strength / i_strength

    return EIBalance(
        ratio=ratio,
        excitatory_strength=e_strength,
        inhibitory_strength=i_strength,
        is_balanced=0.8 <= ratio <= 1.2,
    )


def adjust_ei_ratio(
    knm: FloatArray,
    excitatory_indices: list[int],
    inhibitory_indices: list[int],
    target_ratio: float = 1.0,
) -> FloatArray:
    """Scale inhibitory coupling to achieve target E/I ratio.

    Returns modified knm with inhibitory rows scaled so that
    E_strength / I_strength ≈ target_ratio.
    """
    knm = _validate_knm(knm)
    target_ratio = _validate_target_ratio(target_ratio)
    n = knm.shape[0]
    excitatory_indices = _validate_indices(excitatory_indices, n, "excitatory")
    inhibitory_indices = _validate_indices(inhibitory_indices, n, "inhibitory")

    if _HAS_RUST:
        k_flat = np.ascontiguousarray(knm.ravel())
        e_arr = np.array(excitatory_indices, dtype=np.int64)
        i_arr = np.array(inhibitory_indices, dtype=np.int64)
        result_flat: FloatArray = np.asarray(
            _rust_adjust(k_flat, n, e_arr, i_arr, target_ratio),
        )
        return result_flat.reshape(n, n)

    balance = compute_ei_balance(knm, excitatory_indices, inhibitory_indices)
    if balance.inhibitory_strength < 1e-15 or balance.excitatory_strength < 1e-15:
        return knm.copy()

    current_ratio = balance.ratio
    if abs(current_ratio - target_ratio) < 1e-10:
        return knm.copy()

    # Scale inhibitory rows: I_new = I_old * (current_ratio / target_ratio)
    scale = current_ratio / target_ratio
    result: FloatArray = knm.copy()
    for idx in inhibitory_indices:
        result[idx, :] *= scale
    return result

File: spo-kernel/crates/spo-engine/src/ei_balance.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Phase Orchestrator — Excitatory/Inhibitory balance

//! E/I balance computation for Kuramoto coupling matrices.
//!
//! Kuroki & Mizuseki 2025, Neural Computation — E/I balance is the
//! critical parameter for synchronisation, not K or D.

/// E/I balance result.
///
/// `excitatory_strength` / `inhibitory_strength` aggregate the mean coupling
/// from each source group over all targets. The four `*_to_*` block means
/// resolve this into the directed interaction-type strengths (source group →
/// target group) that Kuroki & Mizuseki 2025 identify as the control
/// parameters of the synchronised / bistable / desynchronised regimes.
pub struct EIBalanceResult {
    pub ratio: f64,
    pub excitatory_strength: f64,
    pub inhibitory_strength: f64,
    pub is_balanced: bool,
    pub e_to_e: f64,
    pub e_to_i: f64,
    pub i_to_e: f64,
    pub i_to_i: f64,
}

/// Mean of `knm[source, target]` over `source ∈ rows`, `target ∈ cols`.
/// Out-of-range indices are skipped; an empty block returns `0.0`.
fn block_mean(knm_flat: &[f64], n: usize, rows: &[usize], cols: &[usize]) -> f64 {
    let mut sum = 0.0;
    let mut count = 0usize;
    for &i in rows {
        if i >= n {
            continue;
        }
        for &j in cols {
            if j >= n {
                continue;
            }
            sum += knm_flat[i * n + j];
            count += 1;
        }
    }
    if count > 0 {
        sum / count as f64
    } else {
        0.0
    }
}

/// Compute E/I balance from coupling matrix and layer typing.
///
/// ratio > 1: excitation-dominated (hypersynchrony risk)
/// ratio < 1: inhibition-dominated (desynchronisation risk)
/// ratio ≈ 1: balanced (optimal for metastability)
///
/// # Arguments
/// * `knm_flat` — (N×N) row-major coupling matrix
/// * `n` — number of oscillators
/// * `excitatory_indices` — indices of excitatory oscillators
/// * `inhibitory_indices` — indices of inhibitory oscillators
#[must_use]
pub fn compute_ei_balance(
    knm_flat: &[f64],
    n: usize,
    excitatory_indices: &[usize],
    inhibitory_indices: &[usize],
) -> EIBalanceResult {
    let e_strength = if excitatory_indices.is_empty() {
        0.0
    } else {
        let mut sum = 0.0;
        let mut count = 0usize;
        for &i in excitatory_indices {
            if i < n {
                for j in 0..n {
                    sum += knm_flat[i * n + j];
                    count += 1;
                }
            }
        }
        if count > 0 {
            sum / count as f64
        } else {
            0.0
        }
    };

    let i_strength = if inhibitory_indices.is_empty() {
        0.0
    } else {
        let mut sum = 0.0;
        let mut count = 0usize;
        for &i in inhibitory_indices {
            if i < n {
                for j in 0..n {
                    sum += knm_flat[i * n + j];
                    count += 1;
                }
            }
        }
        if count > 0 {
            sum / count as f64
        } else {
            0.0
        }
    };

    let ratio = if i_strength.abs() < 1e-15 {
        if e_strength > 0.0 {
            f64::INFINITY
        } else {
            1.0
        }
    } else {
        e_strength / i_strength
    };

    EIBalanceResult {
        ratio,
        excitatory_strength: e_strength,
        inhibitory_strength: i_strength,
        is_balanced: (0.8..=1.2).contains(&ratio),
        e_to_e: block_mean(knm_flat, n, excitatory_indices, excitatory_indices),
        e_to_i: block_mean(knm_flat, n, excitatory_indices, inhibitory_indices),
        i_to_e: block_mean(knm_flat, n, inhibitory_indices, excitatory_indices),
        i_to_i: block_mean(knm_flat, n, inhibitory_indices, inhibitory_indices),
    }
}

/// Scale inhibitory coupling to achieve target E/I ratio.
///
/// Returns modified knm (flat) with inhibitory rows scaled.
#[must_use]
pub fn adjust_ei_ratio(
    knm_flat: &[f64],
    n: usize,
    excitatory_indices: &[usize],
    inhibitory_indices: &[usize],
    target_ratio: f64,
) -> Vec<f64> {
    let balance = compute_ei_balance(knm_flat, n, excitatory_indices, inhibitory_indices);

    if balance.inhibitory_strength.abs() < 1e-15 || balance.excitatory_strength.abs() < 1e-15 {
        return knm_flat.to_vec();
    }
    if (balance.ratio - target_ratio).abs() < 1e-10 {
        return knm_flat.to_vec();
    }

    let scale = balance.ratio / target_ratio;
    let mut result = knm_flat.to_vec();
    for &idx in inhibitory_indices {
        if idx < n {
            for j in 0..n {
                result[idx * n + j] *= scale;
            }
        }
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_balanced_ratio() {
        // Equal E and I strength → ratio = 1.0
        let n = 4;
        let knm = vec![1.0; n * n];
        let e_idx = vec![0, 1];
        let i_idx = vec![2, 3];
        let result = compute_ei_balance(&knm, n, &e_idx, &i_idx);
        assert!((result.ratio - 1.0).abs() < 1e-10);
        assert!(result.is_balanced);
    }

    #[test]
    fn test_interaction_type_breakdown() {
        // 4 oscillators, E = {0, 1}, I = {2, 3}. Each block has a distinct
        // constant value so the directed means are exactly recoverable.
        let n = 4;
        let mut knm = vec![0.0; n * n];
        for &i in &[0usize, 1] {
            for &j in &[0usize, 1] {
                knm[i * n + j] = 2.0; // E→E
            }
            for &j in &[2usize, 3] {
                knm[i * n + j] = 0.5; // E→I
            }
        }
        for &i in &[2usize, 3] {
            for &j in &[0usize, 1] {
                knm[i * n + j] = 1.5; // I→E
            }
            for &j in &[2usize, 3] {
                knm[i * n + j] = 3.0; // I→I
            }
        }
        let r = compute_ei_balance(&knm, n, &[0, 1], &[2, 3]);
        assert!((r.e_to_e - 2.0).abs() < 1e-12);
        assert!((r.e_to_i - 0.5).abs() < 1e-12);
        assert!((r.i_to_e - 1.5).abs() < 1e-12);
        assert!((r.i_to_i - 3.0).abs() < 1e-12);
        // Aggregate excitatory strength = mean over E rows (all targets) =
        // (2.0 + 0.5) / 2 = 1.25; inhibitory = (1.5 + 3.0) / 2 = 2.25.
        assert!((r.excitatory_strength - 1.25).abs() < 1e-12);
        assert!((r.inhibitory_strength - 2.25).abs() < 1e-12);
    }

    #[test]
    fn test_excitation_dominated() {
        let n = 4;
        let mut knm = vec![1.0; n * n];
        // Make excitatory rows stronger
        for j in 0..n {
            knm[0 * n + j] = 3.0;
            knm[1 * n + j] = 3.0;
        }
        let result = compute_ei_balance(&knm, n, &[0, 1], &[2, 3]);
        assert!(
            result.ratio > 1.0,
            "should be excitation-dominated, got {}",
            result.ratio
        );
        assert!(!result.is_balanced);
    }

    #[test]
    fn test_no_inhibitory() {
        let n = 3;
        let knm = vec![1.0; n * n];
        let result = compute_ei_balance(&knm, n, &[0, 1, 2], &[]);
        assert_eq!(result.inhibitory_strength, 0.0);
        assert_eq!(result.ratio, f64::INFINITY);
    }

    #[test]
    fn test_no_excitatory() {
        let n = 3;
        let knm = vec![1.0; n * n];
        let result = compute_ei_balance(&knm, n, &[], &[0, 1, 2]);
        assert_eq!(result.excitatory_strength, 0.0);
        // e_strength = 0, i_strength > 0 → ratio = 0/i = 0
        assert_eq!(result.ratio, 0.0);
    }

    #[test]
    fn test_adjust_ratio_to_target() {
        let n = 4;
        let mut knm = vec![1.0; n * n];
        // E rows twice as strong
        for j in 0..n {
            knm[0 * n + j] = 2.0;
            knm[1 * n + j] = 2.0;
        }
        let adjusted = adjust_ei_ratio(&knm, n, &[0, 1], &[2, 3], 1.0);
        let new_balance = compute_ei_balance(&adjusted, n, &[0, 1], &[2, 3]);
        assert!(
            (new_balance.ratio - 1.0).abs() < 0.1,
            "should be near 1.0 after adjustment, got {}",
            new_balance.ratio
        );
    }

    #[test]
    fn test_adjust_no_change_when_balanced() {
        let n = 3;
        let knm = vec![1.0; n * n];
        let adjusted = adjust_ei_ratio(&knm, n, &[0], &[1, 2], 1.0);
        assert_eq!(adjusted, knm);
    }

    #[test]
    fn test_empty_coupling() {
        let result = compute_ei_balance(&[], 0, &[], &[]);
        assert_eq!(result.ratio, 1.0);
        assert_eq!(result.excitatory_strength, 0.0);
    }

    #[test]
    fn test_out_of_bounds_indices_ignored() {
        let n = 3;
        let knm = vec![1.0; n * n];
        let result = compute_ei_balance(&knm, n, &[0, 100], &[1]);
        // Index 100 is out of bounds, should be skipped
        assert!(result.excitatory_strength > 0.0);
    }
}
